### PR TITLE
Mariana-Pipes-MediaRanker

### PR DESCRIPTION
**Pull request by lmarianarp19**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a model method call sort_all that receives a parameter p and sort it by the number of votes. The first element will be the instance with the bigest number of votes. 
Describe how you approached testing that model method. What edge cases did you come up with? | I tried to create a new work and then vote for this new work. Then I wanted to check if this element was the first element returned for my method. This doesn't work, I couldn't create the necessary instances to test it. 
Describe an edge case test you wrote for a controller | For the index method in the works controller, I test if it returns a success status when there is no works to show.  
What are `session` and `flash`? What is the difference between them? | They are hashes that are used to display messages to the user and store temporary information. The flash is created just for one request while the session is created for a longer period of time, for example while one user is log in.
Describe a controller filter you wrote. |
What was one thing that you gained more clarity on through this assignment? | How the forms works.
What is the Heroku URL of your deployed application | https://marianamediaranker.herokuapp.com/works I couldn't get my data base to run in Heroku.
Do you have any recommendations on how we could improve this project for the next cohort? | It was a lot of requirements. We had to focus in many new concepts  like testing, frameworks, flash, sessions, validations... It should be shorter or we should have more time to finish

### CODE REVIEW

File: app/views/home/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+<h3> Most voted </h3>
+<%= most = @works.sort_all(@works).first.title %>
+ <% collections = [@movies, @books, @albums] %>
+<ul>
+  <% collections.each do |category| %>
+
+
+  <% items = @works.sort_all(category)[0..9] %>
```
Comment (droberts-sea) [inline app/views/home/index.html.erb:9]: This is an example of doing work in the view. Doing this work (sorting, finding the top 10) should be a model method. That method should be called from the controller, and the results stored in `@movies`, `@books` and `@albums`, so that those are ready to go by the time they get here.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      flash.now[:message] = "Failed to create work"
+      flash.now[:details] = @work.errors.messages
+      render :new, status: :bad_request
+    end
+  end
+  # @work = Work.create(title: params[:work][:title], category: params[:work][:category], creator: params[:work][:creator], year: params[:work][:year], description: params[:work][:description])
+
+
+  def show
+    find_work_by_params_id
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:30]: Since you call this method from several actions, it might be wise to make it a controller filter.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | You do a good job using it in `application.html.erb`, but I would like to see more sectioning elements used to divide up individual views.
Errors are reported to the user | yes
Business logic lives in the models | some - I like your `sort_all` method, but I would like you to go ever farther. Something like `Work.top_media` or `Work.top_ten(category)` would mean there's even less work to be done in the view, as well as providing an opportunity to test that functionality.
Models are thoroughly tested, including relations, validations and any custom logic | tests on relations and validations are good, would like to see testing on custom methods, especially as you use them more in the future
Controllers are thoroughly tested | some - `WorksController` has a few test cases but is missing many, and `UsersController` and `VotesController` are completely untested. Make sure you take time to practice this on bEtsy.
**Wave 1 - Media** |
Splash page shows the three media categories | yes - doesn't work if a category has no works
Basic CRUD operations on media are present and functional | mostly - no way to delete a book
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | no
**Wave 3 - Styling** |
Foundation is used appropriately | some - make sure you make an opportunity to practice layouts with Foundation during bEtsy
Look and feel is similar to the original | some
**Overall** |

This is a good start! You were able to get most of the core functionality of the site working, which is impressive.

There are two main places I see room for improvement. The first is code organization: moving as much of the work that's done into the model as possible, and keeping logic out of the views.

The second is around testing, particularly controller testing. Learning to test well is an important part of learning to be an engineer, because it means you can think about your code from the outside. This allows you to anticipate potential problems and pitfalls before they happen, and to problem-solve in an intentional manner rather than fiddling with your code until it works.

That being said, this is a large, difficult project, and we definitely don't expect everything to be perfect. It seems like your Rails learning continues to come along at a good pace, and that's what's important. Keep up the hard work!
